lib/datadog/di/serializer.rb

```diff
@@ -92,12 +102,25 @@ def serialize_value(value, name: nil, depth: settings.dynamic_instrumentation.ma
         when Integer, Float, TrueClass, FalseClass
           serialized.update(value: value.to_s)
         when String, Symbol
-          value = value.to_s
+          need_dup = false
+          value = case value
+          when String
+            # This is the only place where we duplicate the value, currently.
+            # All other values are immutable primitives (e.g. numbers).
+            # However, do not duplicate if the string is frozen, or if
+            # it is later truncated.
+            need_dup = !value.frozen?
+            value
```

Comment:
Is the util `Datadog::Core::Utils::SafeDup.frozen_or_dup` what you are looking for here?